pkg/cosigner/image_signer.go

```diff
@@ -0,0 +1,91 @@
+package cosigner
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/BurntSushi/toml"
+	"github.com/buildpacks/lifecycle"
+	"github.com/sigstore/cosign/cmd/cosign/cli"
+)
+
+type ImageSigner struct {
+	Logger *log.Logger
+}
+
+var (
+	cliSignCmd     = cli.SignCmd
+	secretLocation = "/var/build-secrets"
+)
+
+func NewImageSigner(logger *log.Logger) *ImageSigner {
+	return &ImageSigner{
+		Logger: logger,
+	}
+}
+
+func (s *ImageSigner) Sign(reportFilePath string, annotations map[string]interface{}) error {
+	var report lifecycle.ExportReport
+	_, err := toml.DecodeFile(reportFilePath, &report)
+	if err != nil {
+		return fmt.Errorf("toml decode: %v", err)
+	}
+
+	if len(report.Image.Tags) < 1 {
+		s.Logger.Println("no image tag to sign")
+		return nil
+	}
+
+	cosignFolders := findCosignFolders(secretLocation)
+	if len(cosignFolders) == 0 {
+		s.Logger.Println("no keys found for cosign signing")
+		return nil
+	}
+
+	refImage := report.Image.Tags[0]
+
+	ctx := context.Background()
+	for _, cosignFolder := range cosignFolders {
+		cosignKeyFile := fmt.Sprintf("%s/cosign.key", cosignFolder)
+		cosignPasswordFile := fmt.Sprintf("%s/cosign.password", cosignFolder)
+
+		ko := cli.KeyOpts{KeyRef: cosignKeyFile, PassFunc: func(bool) ([]byte, error) {
+			content, err := ioutil.ReadFile(cosignPasswordFile)
+			if err != nil {
+				return []byte(""), nil
+			}
+
+			return content, nil
+		}}
+
+		if err := cliSignCmd(ctx, ko, annotations, refImage, "", true, "", false, false); err != nil {
+			return fmt.Errorf("unable to sign image with %s: %v", cosignKeyFile, err)
+		}
+	}
+
+	return nil
+}
+
+func findCosignFolders(dir string) []string {
```

Comment:
Can we just mount all cosign creds at their own path and then do a directory list to avoid needing to creatively find the cosign secrets?

Reply:
I've adjusted the buildPod to use a different mountPath when it is a cosign secret.
For finding the directory, I still have to loop through the folder at least and store secret folder names which is leveraged for COSIGN_REPOSITORY and COSIGN_DOCKER_MEDIA_TYPES information. Then using the folder name, I believe I still have to reconstruct the folder path to access the potential "cosign.key" and "cosign.password"

Reply:
Can't we just assume that each directory in `/var/builder/cosign/` properly provides cosign.key and cosign.password?

Reply:
I am assuming as such.

I modified the findCosignFolders function to return the list of `<cosign-secret-name>` where it looks into `/var/builder/cosign/<cosign-secret-name>`.

I could change this to be a walk function such that the signing occurs while it is recursing through the directories. The only requirement was that i needed the `cosign-secret-name` because I was utilizing that to access annotation values in the passed in maps for cosign repository and cosign docker media types.